respy/method_of_simulated_moments.py

```diff
@@ -0,0 +1,327 @@
+"""Estimate models with the method of simulated moments (MSM).
+
+The method of simulated moments is developed by [1], [2], and [3] and an estimation
+technique where the distance between the moments of the actual data and the moments
+implied by the model parameters is minimized.
+
+References
+----------
+
+.. [1] McFadden, D. (1989). A method of simulated moments for estimation of discrete
+       response models without numerical integration. Econometrica: Journal of the
+       Econometric Society, 995-1026.
+.. [2] Lee, B. S., & Ingram, B. F. (1991). Simulation estimation of time-series models.
+       Journal of Econometrics, 47(2-3), 197-205.
+.. [3] Duffie, D., & Singleton, K. (1993). Simulated Moments Estimation of Markov Models
+       of Asset Prices. Econometrica, 61(4), 929-952.
+"""
+import copy
+import functools
+
+import numpy as np
+import pandas as pd
+
+from respy.simulate import get_simulate_func
+
+
+def _harmonize_input(arg):
+    """ Harmonizes different types of inputs by turning all inputs into lists.
+    - pandas.DataFrames/Series and callable functions will turn into list
+        containing a single item (i.e. the input).
+    - Dictionaries will be sorted according to keys and then turn into a list
+        containing the dictionary entries.
+    """
+    # Convert single DataFrames, Series or function into list containing one
+    # item.
+    if isinstance(arg, (pd.DataFrame, pd.Series)) | callable(arg):
+        arg = [arg]
+
+    # Sort dictionary accoding to keys and turn into list.
+    if isinstance(arg, dict):
+        temp = []
+        for key in sorted(arg.keys()):
+            temp.append(arg[key])
+        arg = temp
+
+    # Leave lists as is.
+    if isinstance(arg, list):
+        pass
+    else:
+        raise TypeError(
+            "Function only accepts lists, dictionaries,"
+            " functions, Series and DataFrames as inputs."
+        )
+
+    return arg
+
+
+def _calculate_moments(data, calc_moments):
+    """ Uses functions specified in calc_moments to calculate moments for the
+    data.
+    """
+    data = data.copy()
+    moments = []
+    for func in calc_moments:
+        moments.append(func(data))
+
+    return moments
+
+
+def _reindex_simulated_moments(empirical_moments, simulated_moments):
+    """ Reindexes data in simulated_moments to match index of
+    empirical_moments.
+    """
+    simulated_moments_reindexed = []
+
+    for emp_mom, sim_mom in zip(empirical_moments, simulated_moments):
+        col = emp_mom.columns
+        sim_mom = sim_mom.reindex(columns=col)
+
+        simulated_moments_reindexed.append(sim_mom)
+
+    return simulated_moments_reindexed
+
+
+def _replace_missing_values(simulated_moments, replace_nans):
+    """ Replaces missing values in simulated_moments according to replacement
+    functions specified in replace_nans.
+    """
+    moments = []
+
+    for mom, func in zip(simulated_moments, replace_nans):
```

Comment:
I do not think ``zip()`` works when ``simulated_moments`` has two elements and ``replace_nans`` just one. Then, the ``zip()`` is shortened to just one iteration. If so, we need to bring all arguments to the same length. Can you confirm?

Reply:
Yes, this is true. Currently the user would need to make sure that all lists of inputs have the same length i.e. if there are two elements in `simulated_moments` and the user wants to use the same `replacement_function` for both of them, they would need to define `replace_nans` as `replace_nans  = [replacement_function, replacement_function]`.  We could just fill in the missing functions or shorten  `simulated_moments`  if not enough replacement functions are defined, but maybe an error message would be a better idea to prevent replacement functions being used on the wrong set of moments because the inputs are not defined unambiguously?

Reply:
I am in favor of your first proposal to duplicate the replacement function internally if there is one replacement function and multiple empirical moments. We can do that right after the arguments are harmonized. Otherwise, raise an error. Extending from one to n is very intuitive.

Reply:
Okay, I will do so then. I would suggest adding this to `_replace_missing_values` though since this is the only function that uses `replace_nans` anyway. That will avoid cluttering the code in other places. Unless there is something I am missing?

Reply:
Sounds great :).